### ISSUE TITLE
TASK: determine feasibility of llvm18/clang18 removal wrt python3 JIT

### ISSUE THREAD

**Issue by yselkowitz** [fedora-eln/eln#207]
### What does the ELN SIG need to do?
The new JIT in Python 3.13 has clang as a build (but not runtime) dependency.  Unfortunately, the code is locked to a particular version; 3.13 uses llvm/clang 18, main (the future 3.14) uses 19.  Therefore, a clang18 compat package is needed to build the JIT in Fedora 41 (which normally uses llvm/clang 19).  On Fedora this isn't a big deal, but on ELN this is pulling in llvm18/clang18 into the buildroot (all other consumers have already been ported to the current 19).

For RHEL though this seems untenable.  The default python3.Y package must remain buildable for the entire lifetime of RHEL, but llvm is continuously updated to newer versions.  RHEL 10 has Python 3.12 as default, so there is no precedence yet, but this will come up should any additional versions be provided in parallel (albeit with shorter scope).

Therefore, we need clarification on these points:

- Should the JIT be enabled in RHEL/ELN builds of python 3.13+?
- If so, will a llvm compat version be kept in the buildroot for as long as necessary to continue building the JIT?
- Or, will support for newer llvm versions be backported?  Can that even be done safely, particularly over such a long span?

/cc @hroncok

**Comment by yselkowitz** [comment]
FTR, the changes to port the JIT from 18 to 19 are extensive: https://github.com/python/cpython/pull/125499

**Comment by hroncok** [comment]
Thanks for asking the question, this is definitely something that needs to be solved.

---

>  The default python3.Y package must remain buildable for the entire lifetime of RHEL, but llvm is continuously updated to newer versions. 

That's definitely something we need to consider when we decide whether or not c11s will have Python JIT enabled. Or when we'll add Python 3.14 to c9s/c10s.

> Should the JIT be enabled in RHEL/ELN builds of python 3.13+?

The CPython JIT is experimental, so it definitively isn't "enterprise material". At this point, I would not enable it in RHEL.

On the other hand, ELN is a place to experiment for future RHEL. Chances are that the JIT will become more stable in time for RHEL 11 (which will very likely contain Python 3.15/[3.26](https://peps.python.org/pep-2026/)).

So I'd like to keep this enabled in ELN for now (if it's not a big deal) and revisit this in 2 years. I would certainly not want to pull some old LLVM version to RHEL-proper, but but having it in ELN "for now" is probably OK.

> If so, will a llvm compat version be kept in the buildroot for as long as necessary to continue building the JIT?

If so, then yes.

> Will support for newer llvm versions be backported? 

Time will tell. No idea yet.

**Comment by hroncok** [comment]
Upstreams considers supporting ranges of LLVM versions: https://discuss.python.org/t/llvm-19-for-jit-on-3-13-maybe-19-for-3-14/73425/2

**Comment by yselkowitz** [comment]
FTR, [LLVM 20 is planned](https://fedoraproject.org/wiki/Changes/LLVM-20) for F42, while Python will stay on 3.13.  That means, barring any other changes, LLVM 18 will need to remain in for the duration of F42, as well as in ELN until 3.14.  Hopefully 3.14 will be compatible with 20 and 21 in due course, otherwise Python (with its yearly cadence) will never keep up with LLVM (with its semi-yearly cadence).

As much as I wish to drop llvm/clang 18 from ELN, I guess there is nothing to do until 3.14 lands, at which point we can revisit this.

**Comment by hroncok** [comment]
See also this PEP draft: https://peps.python.org/pep-0774/

tl;dr we might be able to conditionalize the dependency and keep JIT even in EL(N)

**Comment by yselkowitz** [comment]
But FWIW the proposal only mentions Tier 1 platforms, which IIUC does not include ppc64le/s390x, unless you would just pregenerate those and keep them in dist-git?

**Comment by hroncok** [comment]
That's still to be determined but keeping them in distgit is an option.

**Comment by hroncok** [comment]
We currently only build the JIT on aarch64 and x86_64 anyway.

**Comment by yselkowitz** [comment]
Reopening as a known issue for tracking the eventual removal of llvm18/clang18.

**Comment by yselkowitz** [comment]
FTR, llvm-20 just landed in rawhide/F43 and is on its way to ELN and F42, making llvm18 N-2 and (probably) the Tools team a bit more anxious to remove it.

**Comment by hroncok** [comment]
[PEP 774 was deferred](https://discuss.python.org/t/pep-774-removing-the-llvm-requirement-for-jit-builds/78548/34).

> This is the kind of build change we expect to be taken when expecting the JIT to no longer be an experimental build feature. But it is too soon to consider the JIT not experimental so we feel it is too soon to do this.

So in c11s, we'll end up either:

 - disabling the JIT as it is way to experimental to be included in an enterprise OS
 - using pre-built stencils if JIT is no longer experimental and PEP 774 is approved
 - (unlikely) using downstream pre-built stencils if there is significant demand to ship the JIT even when upstream says it's experimental

All of the options will remove the llvm/clang dependency.

Either way, I'd like to keep the status quo in ELN for now, for experiment purposes. However, if there is significant pressure (from llvm maintainers or ELN SIG), we can disable it.

**Comment by hroncok** [comment]
> FTR, llvm-20 just landed in rawhide/F43 and is on its way to ELN and F42, making llvm18 N-2 and (probably) the Tools team a bit more anxious to remove it.

We will update Python to 3.14 in May/June. That will allow the removal od llvm18, as 3.14 uses llvm19.

**Comment by yselkowitz** [comment]
This just doesn't seem ready, as it's not keeping up (in fact, falling behind) LLVM's pace, as I was afraid would happen.  If 3.14 is still going to require llvm19, then by this time next year it will be on N-3 (!) when llvm is bumped to 22 in rawhide and ELN.

Therefore, please disable the JIT in RHEL/ELN builds of 3.14 prior to its becoming the default Python in rawhide.  We can reassess for 3.15/3.26 next year, as that is the version that would be branched for RHEL 11.

**Comment by hroncok** [comment]
>  If 3.14 is still going to require llvm19, then by this time next year...

We don't know that.

> ...then by this time next year it will be on N-3 (!) when llvm is bumped to 22 in rawhide and ELN.

Next year, we will be updating to 3.15.

> Therefore, please disable the JIT in RHEL/ELN builds of 3.14 prior to its becoming the default Python in rawhide.

I can certainly do that, but I'd rather keep building the JIT in ELN to test it in case we end up needing it in 3.15. Where else to experiment with future EL features than in ELN?

**Comment by hroncok** [comment]
Use prebuilt JIT stencils on platforms without LLVM 19: https://src.fedoraproject.org/rpms/python3.14/pull-request/40 (work in progress)

**Comment by hroncok** [comment]
The work got stalled, so let's disable JIT on RHEL in the meantime: https://src.fedoraproject.org/rpms/python3.14/pull-request/76

This will get into ELN when Python 3.14 becomes the main Python. I can backport this to Python 3.13 as well, but I don't think this necessarily needs to land right now.

**Comment by yselkowitz** [comment]
3.14 is around the corner, so that's fine.

**Comment by yselkowitz** [comment]
This will land with the 3.14 mass rebuild.

**Comment by hroncok** [comment]
FWIW, this has been resolved for Python 3.15 in https://src.fedoraproject.org/rpms/python3.15/pull-request/17

As a result:

 - Python in RHEL/ELN does not depend on LLVM
 - Python in RHEL/ELN still has JIT

On Python 3.14 we will either keep the JIT disabled or backport this, but that will be irrelevant for c11s.
